# Task
preprocess dataset

Here is all the data you need:
"spam (2).csv"

## Data loading

### Subtask:
Load the dataset "spam (2).csv" into a pandas DataFrame.


**Reasoning**:
Load the dataset "spam (2).csv" into a pandas DataFrame and display its first few rows.



In [ ]:
import pandas as pd

try:
    df = pd.read_csv('spam (2).csv', encoding='latin-1')
    display(df.head())
    display(df.info())
except FileNotFoundError:
    print("Error: 'spam (2).csv' not found. Please ensure the file exists in the current directory.")
except Exception as e:
    print(f"An error occurred: {e}")

v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   v1          5572 non-null   object
 1   v2          5572 non-null   object
 2   Unnamed: 2  50 non-null     object
 3   Unnamed: 3  12 non-null     object
 4   Unnamed: 4  6 non-null      object
dtypes: object(5)
memory usage: 217.8+ KB


None

## Data exploration

### Subtask:
Explore the loaded dataset to understand its structure, identify potential issues, and inform subsequent cleaning and preparation steps.


**Reasoning**:
Analyze the distribution of features, identify potential inconsistencies or irrelevant columns, and summarize the findings.



In [ ]:
# Analyze the distribution of features
print(df['v1'].value_counts())  # Frequency of categories in 'v1'
print(df['v2'].value_counts().head(10)) # Top 10 most frequent messages in 'v2'

# Identify potential inconsistencies or irrelevant columns
print(df.isnull().sum()) # Check for missing values again

# Analyze 'Unnamed' columns
for col in ['Unnamed: 2', 'Unnamed: 3', 'Unnamed: 4']:
    print(f"\nAnalysis of column '{col}':")
    print(df[col].value_counts(dropna=False)) # Include NaN counts

# Summarize findings
print("\nSummary of Findings:")
print("1. Data Types and Missing Values:")
print(" - 'v1' and 'v2' have no missing values.")
print(" - 'Unnamed: 2', 'Unnamed: 3', and 'Unnamed: 4' have a large number of missing values.")
print(" - All columns are of object dtype, which might need further investigation, especially for 'v2'.")

print("\n2. Distribution of Features:")
print(" - 'v1' column shows the distribution of 'ham' and 'spam' messages.")
print(" - 'v2' column shows a wide variety of text messages, with some messages appearing multiple times.")

print("\n3. Potential Inconsistencies or Irrelevant Columns:")
print(" - 'Unnamed: 2', 'Unnamed: 3', and 'Unnamed: 4' contain mostly null values and appear irrelevant for the task.")
print(" - The 'v2' column should be examined for inconsistencies in text format.")

print("\n4. Columns with mostly null values:")
print(" - 'Unnamed: 2', 'Unnamed: 3', and 'Unnamed: 4' columns are candidates for removal due to high null value counts.")
print(" - The usefulness of these columns is questionable and they might be dropped in the next steps.")

v1
ham     4825
spam     747
Name: count, dtype: int64
v2
Sorry, I'll call later                                                                                                                                                                 30
I cant pick the phone right now. Pls send a message                                                                                                                                    12
Ok...                                                                                                                                                                                  10
Wen ur lovable bcums angry wid u, dnt take it seriously.. Coz being angry is d most childish n true way of showing deep affection, care n luv!.. kettoda manda... Have nice day da.     4
Okie                                                                                                                                                                                    4
7 wonders in

## Data cleaning

### Subtask:
Clean the DataFrame `df` by removing irrelevant columns and handling duplicate rows.


**Reasoning**:
Drop the irrelevant columns and remove duplicate rows from the DataFrame `df` as instructed, then display the first 5 rows and the info of the cleaned DataFrame.



In [ ]:
# Drop irrelevant columns
df = df.drop(columns=['Unnamed: 2', 'Unnamed: 3', 'Unnamed: 4'])

# Remove duplicate rows
df = df.drop_duplicates(keep='first')

# Display first 5 rows and info
display(df.head())
display(df.info())

v1                                                 v2
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

<class 'pandas.core.frame.DataFrame'>
Index: 5169 entries, 0 to 5571
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   v1      5169 non-null   object
 1   v2      5169 non-null   object
dtypes: object(2)
memory usage: 121.1+ KB


None

## Data preparation

### Subtask:
Convert the categorical feature 'v1' in the DataFrame `df` into numerical representations using label encoding.


**Reasoning**:
Convert the categorical feature 'v1' to numerical labels using LabelEncoder.



In [ ]:
from sklearn.preprocessing import LabelEncoder

# Create a LabelEncoder object
label_encoder = LabelEncoder()

# Fit the encoder on the 'v1' column
label_encoder.fit(df['v1'])

# Transform the 'v1' column and overwrite the original column
df['v1'] = label_encoder.transform(df['v1'])

# Display the first few rows and info to verify
display(df.head())
display(df.info())

v1                                                 v2
0   0  Go until jurong point, crazy.. Available only ...
1   0                      Ok lar... Joking wif u oni...
2   1  Free entry in 2 a wkly comp to win FA Cup fina...
3   0  U dun say so early hor... U c already then say...
4   0  Nah I don't think he goes to usf, he lives aro...

<class 'pandas.core.frame.DataFrame'>
Index: 5169 entries, 0 to 5571
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   v1      5169 non-null   int64 
 1   v2      5169 non-null   object
dtypes: int64(1), object(1)
memory usage: 121.1+ KB


None

## Summary:

### 1. Q&A

No questions were explicitly asked in the provided data analysis task.  However, the steps implicitly address questions like:

*   How can we prepare the "spam (2).csv" dataset for analysis?
*   What are the key characteristics of the dataset?
*   Are there any irrelevant or problematic columns?
*   How can we handle categorical variables like the 'v1' column?


### 2. Data Analysis Key Findings

*   **Irrelevant Columns Removed:** The 'Unnamed: 2', 'Unnamed: 3', and 'Unnamed: 4' columns were removed due to a high number of missing values (over 5500 out of 5572 rows).
*   **Duplicate Rows Removed:** 403 duplicate rows were removed, reducing the dataset size from 5572 to 5169 entries.
*   **Categorical Feature Encoding:** The 'v1' column (containing 'ham' and 'spam' labels) was successfully converted to numerical labels (0 and 1) using label encoding.


### 3. Insights or Next Steps

*   **Text Preprocessing:** The 'v2' column (text messages) requires further preprocessing, including cleaning (removing punctuation, converting to lowercase), tokenization, and potentially stemming/lemmatization before being used in a machine learning model.
*   **Model Training:**  With the cleaned and prepared data, the next step should be to train a machine learning model to classify messages as 'ham' or 'spam' based on the preprocessed text in the 'v2' column.
